src/thunderbots/software/util/parameter/parameter.h

```diff
@@ -16,118 +29,198 @@
+         * Updates the value of this Parameter with the value from the ROS
+         * Parameter Server
+         */
+        void updateValueFromROSParameterServer()
+        {
+            ros::param::get(getROSParameterPath(), value_);
+        }
+
+        /**
+         * Sets the parameter value in the ROS Parameter Server to the new value.
+         *
+         * @param new_value The new value to be set
+         */
+        void setValueInROSParameterServer(T new_value)
+        {
+	    //dynamic reconfigure takes control of setting params //ros::param::set(getROSParameterPath(), new_value);
+        }
+
+        /**
+         * Returns a reference to the Parameter registry. The registry is a list of
+         * pointers to all the existing Parameters.
+         *
+         * @return An immutable reference to the Parameter registry
+         */
+        static const std::vector<std::unique_ptr<Parameter<T>>>& getRegistry()
+        {
+            return Parameter<T>::getMutableRegistry();
+        }
+
+        /**
+         * Returns a reference to the config struct. The config struct contains
+         * all the current configurations
+         *
+         * @return An immutable reference to the Config struct
+         */
+        static const dynamic_reconfigure::Config& getConfigStruct()
         {
-            p->updateValueFromROSParameterServer();
+            return Parameter<T>::getMutableConfigStruct();
         }
-    }
-
-   private:
-    /**
-     * Returns a mutable reference to the Parameter registry. This is the same as the
-     * above getRegistry() function, except that it returns a mutable reference. We need a
-     * mutable version so that member functions can modify the registry. The function is
-     * private so that external sources cannot modify the registry in unexpected ways.
-     *
-     * @return A mutable reference to the Parameter registry
-     */
-    static std::vector<std::unique_ptr<Parameter>>& getMutableRegistry()
-    {
-        static std::vector<std::unique_ptr<Parameter<T>>> instance;
-        return instance;
-    }
-
-    // Store the value so it can be retrieved without fetching from the server again
-    T value_;
-
-    // The global path to the variable in the ROS Parameter Server
-    std::string ros_parameter_path;
+
+        /**
+         * Registers (adds) a Parameter to the registry. Since the unique pointer is moved
+         * into the registry, the pointer may not be accessed by the caller after this
+         * function has been called.
+	 *
+	 * Also registers params to the static configuration struct used to 
+	 * set parameters
+         *
+         * @param parameter A unique pointer to the Parameter to add. This pointer may not
+         * be accessed by the caller after this function has been called.
+         */
+        static void registerParameter(std::unique_ptr<Parameter<T>> parameter)
+        {   
+
+            if constexpr(std::is_same<T, bool>::value){   
```

Comment:
Instead of using `std::is_same` here, can we template the `registerParameter` function? ie, 
```
static void registerParameter<T>(std::unique_ptr<Parameter<T>> parameter)
```
and then have concrete instances of it for the types that we care about?

Heck, we might be able to get away with doing something like we did for `Play` (ie. a static assert, then maybe a cast?), ie:
```
template <class T>
class TPlayFactory : public PlayFactory
{
    // compile time type checking that T is derived class of Play
    static_assert(std::is_base_of<Play, T>::value, "T must be derived class of Play!");

   public:
    TPlayFactory()
    {
        Play::registerPlay(std::shared_ptr<PlayFactory>(this));
    }

    std::shared_ptr<Play> getInstance() override
    {
        return std::make_shared<T>();
    }
};
```
This would reduce a lot of duplication here, which would be great.
(templating can get a bit confusing, and I'm not positive what I've written about is 100% correct, so feel free to ping me and we can sit down and figure it out if you like)

Reply:
After some consideration, retracting this comment, I don't think this will work here.